ppp/discovery.go

```diff
@@ -14,40 +20,303 @@ const (
-		0x11, // Protocol version 1, packet type 1 (both constants)
-		9,    // Code PADI, aka "hey any PPPoE concentrators out there?"
-		0, 0, // Session ID, not set during discovery
-		0, 4, // Payload length. payload is a TLV array of tags, but we're sending only 1 zero-length tag
-		1, 1, // Tag "Service-Name", aka "which ISP is acceptable to me?
-		0, 0, // Tag value length, 0
-		// No tag value, meaning "any ISP is acceptable"
-	}
+	padiPacket = encodeDiscoveryPacket(&discoveryPacket{
+		Code: pppoePADI,
+		TLV: map[int][]byte{
+			// By convention on single-ISP customer access networks,
+			// the tag is always nil, meaning "don't care," because
+			// there's only one ISP around anyway.
+			pppoeTagServiceName: nil,
+		},
+	})
 	// ethernetBroadcast is the Ethernet broadcast address.
 	ethernetBroadcast = &raw.Addr{
 		HardwareAddr: net.HardwareAddr{0xff, 0xff, 0xff, 0xff, 0xff, 0xff},
 	}
 )
 
 // pppoeDiscovery executes PPPoE discovery and returns a PPPoE session ID.
-func pppoeDiscovery(ifName string) (sessionID uint16, err error) {
+func pppoeDiscovery(ctx context.Context, ifName string) (sessionID int, closer func() error, err error) {
+	conn, err := newDiscoveryConn(ifName)
+	if err != nil {
+		return 0, nil, err
+	}
+	defer func() {
+		if err != nil {
+			conn.Close()
+		}
+	}()
+
+	deadline, hasDeadline := ctx.Deadline()
+
+	var (
+		concentrator net.Addr
+		cookie       []byte
+	)
+
+	// Broadcast PADIs, looking for a PPPoE concentrator.
+	for concentrator == nil && (!hasDeadline || time.Now().Before(deadline)) {
+		// Send a PADI, asking concentrators for a session offer.
+		if err := sendPADI(conn); err != nil {
+			return 0, nil, fmt.Errorf("sending PADI packet: %v", err)
+		}
+
+		padoCtx, cancelPADO := context.WithTimeout(ctx, time.Second)
+		defer cancelPADO()
+		concentrator, cookie, err = readPADO(padoCtx, conn)
+		if err == nil {
+			// We know about a concentrator, move on.
+			break
+		} else if neterr, ok := err.(net.Error); !ok || !neterr.Timeout() {
+			return 0, nil, fmt.Errorf("waiting for PADO: %v", err)
+		}
+		// Timed out waiting for PADO. Loop back around to (maybe) try
+		// again.
+	}
+
+	// Got a concentrator, request a session.
+	for !hasDeadline || time.Now().Before(deadline) {
+		if err := sendPADR(conn, concentrator, cookie); err != nil {
+			return 0, nil, fmt.Errorf("sending PADR packet: %v", err)
+		}
+
+		padsCtx, cancelPADS := context.WithTimeout(ctx, time.Second)
+		defer cancelPADS()
+		sessionID, err = readPADS(padsCtx, conn, concentrator)
+		if err == nil {
+			// We're done!
+			return sessionID, func() error { return sendPADT(conn, concentrator, sessionID) }, nil
+		} else if neterr, ok := err.(net.Error); !ok || !neterr.Timeout() {
+			return 0, nil, fmt.Errorf("waiting for PADS: %v", err)
+		}
+		// Timed out waiting for PADS. Loop back around to (maybe) try
+		// again.
+	}
+
+	// Oops, deadline exceeded :(
+	return 0, nil, errors.New("deadline exceeded during PPPoE Discovery")
```

Comment:
You can't ask any questions of this error value. I'd just return the ctx.Err() in this case?

Reply:
Alternatively, make your own error type implementing net.Error.

Reply:
Returning ctx.Err().